### PR TITLE
ANN: Detect E0728 error (`await` outside `async` function)

### PR DESCRIPTION
**Pull request by dima74**
changelog: Detect [E0728](https://doc.rust-lang.org/error-index.html#E0728) compiler error

### CODE REVIEW

**Review by ortem — APPROVED** [review]

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsExpr.kt [hunk 1]
```diff
@@ -322,19 +322,6 @@ val RsExpr.isInUnsafeContext: Boolean
         return parent is RsBlockExpr || (parent is RsFunction && parent.isActuallyUnsafe)
     }
 
-val RsExpr.isInAsyncContext: Boolean
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/psi/ext/RsExpr.kt:325]: Yep, that was a bug, and so we can qualify this change as a fix. In this case I think we don't have to worry

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt [hunk 2]
```diff
@@ -88,6 +85,18 @@ val PsiElement.isAtLeastEdition2018: Boolean
 val RsElement.isConstantLike: Boolean
     get() = this is RsConstant || (this is RsFieldsOwner && isFieldless)
 
+val RsElement.isInAsyncContext: Boolean
+    get() {
+        for (context in contexts) {
+            when (context) {
+                is RsBlockExpr -> if (context.isAsync) return true
+                is RsFunction -> return context.isAsync
+                is RsLambdaExpr -> return context.isAsync
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt:94]:
```suggestion
                is RsFunctionOrLambda -> return context.isAsync
```

**Comment by dima74** [comment]
bors r=ortem

**Review by Undin — COMMENTED** [review]
It seems the plugin suggests making function async even if it already returns `Future`. I suppose, we shouldn't provide `Make function async` quick-fix in this situation (at least) or provide `Surround with async block`/`Add async to block` quick fixes (or something similar) like we do with `unsafe`

For example, see code from [async book](https://rust-lang.github.io/async-book/03_async_await/01_chapter.html):

```rust
async fn foo() -> u8 { 5 }

fn bar() -> impl Future<Output = u8> {
    let x: u8 = foo().await;
    x + 5
}
```

**Pull request merged**